### ISSUE TITLE
🐛 [Bug]: QueryParser is not parsing comma separated values to slices by default

### ISSUE THREAD

**Issue by lacerdev** [gofiber/fiber#2624]
### Bug Description

Bug inserted into #2560. 

I agree that BodyParser shouldn't be splitting values by comma by default, but I don't agree that the same behavior should be extrapolated to QueryParser. By default query params separated by commas are arrays/slices and I believe that the framework must follow this rule by default and not by enabling a configuration

### How to Reproduce

```bash
curl --location 'http://localhost:5000/api/alerts?level=warning%2Cerror' 
```

```golang
package main

import "github.com/gofiber/fiber/v2"
import "log"

func main() {
  app := fiber.New()

  // Steps to reproduce
	app.Post("/alerts",func(ctx *fiber.Ctx) error {
		type ReqParams struct {
			Level []string `query:"level"`
		}

		var params ReqParams
		ctx.QueryParser(&params)

	    //incorrect member size 1
          // params.Level[0] = "warning,error" 
        
         // correct members should be:  params.Level[0]="warning"      params.Level[1]="error"
		return nil
	})
  log.Fatal(app.Listen(":5000"))
}
```

1. Go to '....'
3. Do '....'
4. See '....'

### Expected Behavior

Query params must be parsed to slices when they came separated with commas

### Fiber Version

v2.49.1

### Code Snippet (optional)

```go
package main

import "github.com/gofiber/fiber/v2"
import "log"

func main() {
  app := fiber.New()

  // Steps to reproduce

  log.Fatal(app.Listen(":3000"))
}
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my problem prior to opening this one.
- [X] I understand that improperly formatted bug reports may be closed without explanation.

**Comment by ReneWerner87** [comment]
@rere61 what do you think about it ?

**Comment by rere61** [comment]
> @rere61 what do you think about it ?

Nice. Great Work.
Thanks for all your help.
GoFiber is my favorite framework, lol.

**Comment by ReneWerner87** [comment]
@rere61 actually meant about the current behavior and the new behavior which is requested in the report

**Comment by rere61** [comment]
> @rere61 actually meant about the current behavior and the new behavior which is requested in the report

Yes, the new behavior is more common compared to others frameworks.
And this issuse is completed solved, please close this issuse. thanks.

**Comment by ReneWerner87** [comment]
> And this issuse is completed solved, please close this issuse. thanks.

@rere61 
this is not your original report, but one from someone else
please take your time and read from top to bottom

**Comment by rere61** [comment]
> > And this issuse is completed solved, please close this issuse. thanks.
> 
> @rere61 this is not your original report, but one from someone else please take your time and read from top to bottom

@ReneWerner87 
Okay. I got it.
Please take a look at my original post.
I write many examples and the comparsion of Go standard HTTP/Gin, others frameworks in others language.

**Comment by lacerdev** [comment]
Because that's what frameworks are for. So we don't have to do the "dirty work".

Go Standard HTTP is a raw implementation, so it's understandable that we must do these transformations manually

Comma-separated query parameters are by convention transformed into arrays/lists/slices of values around the web developer community.

Again, in case of `BodyParser` I agree with this behavior of not parsing comma-separated values into slices, but to the `QueryParser` it should be the default behavior by convention, since query parameters are usually filter parameters

Or, another solution is decide based on param type, if that field was defined as `string` type, treat like string, if defined as `[]string` treat like slice and split values by comma.

**Comment by rere61** [comment]
> Or, another solution is decide based on param type, if that field was defined as `string` type, treat like string, if defined as `[]string` treat like slice and split values by comma

Can you tell me which framework will automatically separate a user's **single** query value, such as input containing a given name and first name(ex:   Hemingway, Ernest Miller), with commas without causing side effects?

**Comment by lacerdev** [comment]
> 

In this specific case the query parameters must be encoded and decoded on server side. But, this is an atypical case, in most cases the comma will be used as a separator and not as an integral part of the string.

It is possible turn the method smarter as below

```golang
if equalFieldType(out, reflect.Slice, k) {
			values := strings.Split(v, ",")
			for i := 0; i < len(values); i++ {
				data[k] = append(data[k], values[i])
			}
		} else {
			data[k] = append(data[k], v)
		}
```
It would be not needed to check if `strings.Contains(v, ",")`. This way if the developer would be considering receiving the value as slice, the framework tries to do `splitting`, if does not contains ",", would just return a slice of 1 position with full string and the code flows naturally


Another option is to enable/disable this behavior as a new parameter in the QueryParser method, instead of a global config.

```golang
// enableSplitting is a second param in QueryParser method, or maybe could be exist a options{} second parameter containing some configs, one of them enableSplitting. And by default this value would be true
if equalFieldType(out, reflect.Slice, k) && enableSplitting {
			values := strings.Split(v, ",")
			for i := 0; i < len(values); i++ {
				data[k] = append(data[k], values[i])
			}
		} else {
			data[k] = append(data[k], v)
		}
```

this way It would not be a _breaking change_ and will not change the behavior for entire application

**Comment by joey1123455** [comment]
> > 
> 
> In this specific case the query parameters must be encoded and decoded on server side. But, this is an atypical case, in most cases the comma will be used as a separator and not as an integral part of the string.
> 
> It is possible turn the method smarter as below
> 
> ```golang
> if equalFieldType(out, reflect.Slice, k) {
> 			values := strings.Split(v, ",")
> 			for i := 0; i < len(values); i++ {
> 				data[k] = append(data[k], values[i])
> 			}
> 		} else {
> 			data[k] = append(data[k], v)
> 		}
> ```
> It would be not needed to check if `strings.Contains(v, ",")`. This way if the developer would be considering receiving the value as slice, the framework tries to do `splitting`, if does not contains ",", would just return a slice of 1 position with full string and the code flows naturally
> 
> 
> Another option is to enable/disable this behavior as a new parameter in the QueryParser method, instead of a global config.
> 
> ```golang
> // enableSplitting is a second param in QueryParser method, or maybe could be exist a options{} second 
> // parameter containing some configs, one of them enableSplitting. And by default this value would be true
> if equalFieldType(out, reflect.Slice, k) && enableSplitting {
> 			values := strings.Split(v, ",")
> 			for i := 0; i < len(values); i++ {
> 				data[k] = append(data[k], values[i])
> 			}
> 		} else {
> 			data[k] = append(data[k], v)
> 		}
> ```
> 
> this way It would not be a _breaking change_ and will not change the behavior for entire application

This isn't a bad idea actually but it would be nice to have it also take into account the global config

**Comment by joey1123455** [comment]
> And if it has two global configs? One for BodyParser and other for QueryParser? Where the QueryParser config would be true by default. And, besides that, could have this second param too, that would be initialized by the global config value, but that could be changed if user has passed as argument...

Hmm or instead by default it would take the global variable if true by default, if its not set then a local variable takes preference.

**Comment by khawkinson** [comment]
> > > And this issuse is completed solved, please close this issuse. thanks.
> > 
> > 
> > @rere61 this is not your original report, but one from someone else please take your time and read from top to bottom
> 
> @ReneWerner87 Okay. I got it. Please take a look at my original post. I write many examples and the comparsion of Go Fiber v.s Go standard HTTP v.s Gin.
> 
> So, my short answer for this request is "No". We can not make GoFiber different than Go net/http, Gin, others framework in Go, all frameworks of all programmin languages in the world.
> 
> And most imporantly, split string by comma is very easy task, why should this feature needed to be built-in framework?

 `ctx.QueryParser()` in v2.46.0 would parse these three query params the same:
`?years=2023,2024`
`?years=2023&years=2024`
`?years[]=2023&years[]=2024`
which would always end up being `[]string{"2023","2024"}`
now with the changes with v2.52.0 the first example ends up being `[]string{"2023,2024"}` while the other two remain the same. 

Are we expecting to have to parse the string ourselves going forward? If this is the case someone should probably update the [docs](https://docs.gofiber.io/api/ctx/#queryparser) because that is claiming otherwise

**Comment by ReneWerner87** [comment]
done with the config setting
[`EnableSplittingOnParsers` 
](https://docs.gofiber.io/api/fiber#enablesplittingonparsers)
